client/components/Editor/plugins/keymap.ts

```diff
@@ -150,14 +152,55 @@ export default (schema) => {
 		);
 	}
 
-	// All but the custom block splitting command in this chain are taken from the default
-	// chain in baseKeymap. We provide our own block splitter that preserves text align
-	// attributes between paragraphs.
+	const customBlockSplitter = splitBlockPreservingAttrs(['textAlign', 'rtl']);
+
+	// This command runs the link input rule (converting emails and urls into links) whenever the
+	// user presses enter. Adding this to our enter handler is a hack to make sure the input rule
+	// behavior still works even though input rules don't work across nodes (see
+	// https://discuss.prosemirror.net/t/trigger-inputrule-on-enter/1118/4 and
+	// https://github.com/ProseMirror/prosemirror-inputrules/pull/6#issuecomment-894107661 for
+	// context)
+	const addLinkCommand = (state: EditorState, dispatch?: Dispatch) => {
+		if (!(state.selection instanceof TextSelection)) {
+			return false;
+		}
+		const $cursor = state.selection.$cursor;
+		if (!$cursor) {
+			return false;
+		}
+		const { nodeBefore } = state.selection.$from;
+		if (!nodeBefore || !nodeBefore.isText) {
+			return false;
+		}
+
+		const match = nodeBefore.text!.match(EMAIL_OR_URI_REGEX);
+		if (!match) {
+			return false;
+		}
+
+		if (dispatch) {
+			// Call our custom split block command, then add the link marks as part of the same
+			// transaction. This should be safe to do without mapping the start and end positions
+			// because the changes caused by splitBlock will all be after the cursor
+			customBlockSplitter(state, (tr) => {
+				const addLinkMark = linkRuleHandler(schema.marks.link, false, tr);
+				const start = $cursor.pos - match[0].length;
+				const end = $cursor.pos;
+				dispatch(addLinkMark(state, match, start, end));
```

Comment:
This is all necessary because we're running this code as one of the possible enter key handlers, and we still want the usual behavior inside a text block of creating a new paragraph when you press enter.

When I first started, I misunderstood how `chainCommands` below worked, and assumed we could add our command at the beginning, and then let the rest of them run as usual. However, the next command isn't invoked if any of the previous ones returns true, and users are explicitly cautioned not to create side effects inside of a command that returns false (https://discuss.prosemirror.net/t/chaincommands-seems-to-be-working-wrong/1321/2). So the only way to retain the normal enter handler behavior with this new command is to combine the link changes into the same transaction as whatever other command we run. Luckily in this case (inside a text block, $cursor selection), I'm pretty sure only `splitBlockPreservingAttrs(['textAlign', 'rtl'])` could possibly run.